Cps/Adding embedded case -  Initialize geometry - 2 of 4

marcnunezc opened the pull request:
Hi,
This is a follow up of #4681
- [x] 1. Adding embedded element
- [ ] 2. Adding initialize geometry script with remeshing
- [ ] 3. Adding embedded wake and embedded compute lift processes
- [ ] 4. Adding custom metric level-set (refine close to leading edge and trailing edge).

In this second PR I'm adding the script needed to initialize a skin geometry, remesh according to it, and solve the potential flow around it **(no wake yet)**:

No remeshing

![image](https://user-images.githubusercontent.com/32136457/57288690-3e1fa500-70ba-11e9-94c1-c5501a6c0f60.png)


Remeshed

![image](https://user-images.githubusercontent.com/32136457/57288635-23e5c700-70ba-11e9-8bd9-fd643dcda677.png)


In order to do so I have a python script that calls the mmg processes from the MeshingApplication (I'm adding @loumalouomega so he can review this too).

I added the changes from #4804 so that I dont have to find the cut elements again in a different process (since they are already found in the modification_distance_process). That's why the flag in the embedded element has changed from BOUNDARY to TO_SPLIT. 

I also added a python test.


Thanks!

marcnunezc committed:
adding initialize geometry script

marcnunezc committed:
Adding potential flow formulation for the embedded case and adding embeddedvariables

marcnunezc committed:
Cleaning up script and adding comments

marcnunezc committed:
Defining apply flags process in C++

marcnunezc committed:
typo

marcnunezc committed:
Using GetVector()

marcnunezc committed:
Defining MoveModelPart function

marcnunezc committed:
Adding update parameters function

marcnunezc committed:
renaming class to refer to the embedded case

marcnunezc committed:
renaming flag process

marcnunezc committed:
removing unused lines and includes

marcnunezc committed:
Adding test for cpp process

marcnunezc committed:
Merge branch 'cps/embedded-pr-1of4' into cps/embedded-pr-2of4

marcnunezc committed:
Updating LEVEL_SET variable name

marcnunezc committed:
Adding a python test for the embedded case with no wake

marcnunezc committed:
Merge remote-tracking branch 'origin/master' into cps/embedded-pr-2of4

marcnunezc committed:
Merge branch 'cps/embedded-hotfix' into cps/embedded-pr-2of4

marcnunezc committed:
cleaning up process

marcnunezc committed:
Adding distance modification process

marcnunezc committed:
using set to zero processes

marcnunezc committed:
updating test after adding distance modification

marcnunezc committed:
Merge branch 'cps/using-normal-from-geometry'

marcnunezc committed:
Merge branch 'master' into cps/embedded-pr-2of4

marcnunezc committed:
removing normal calculation utils call

marcnunezc committed:
Updating freestream velocity name

marcnunezc committed:
Merge remote-tracking branch 'origin/master' into cps/embedded-pr-2of4

marcnunezc committed:
moving loading of skin model part

loumalouomega committed:
Merge branch 'master' into cps/embedded-pr-2of4

marcnunezc committed:
Merge remote-tracking branch 'origin/meshing/flag-to-erase-reset-flag' into cps/embedded-pr-2of4

marcnunezc committed:
Removing custom apply flags process and using the modification distance process flags

marcnunezc committed:
Merge remote-tracking branch 'origin/fluid/to-split-flag-in-distance-modification' into cps/embedded-pr-2of4

marcnunezc committed:
updating test

marcnunezc committed:
Removing custom apply flags call

philbucher commented on applications/CompressiblePotentialFlowApplication/python_scripts/initialize_geometry.py:
```diff
@@ -0,0 +1,252 @@
+        KratosMultiphysics.VariableUtils().SetNonHistoricalVariableToZero(MeshingApplication.METRIC_TENSOR_2D, self.main_model_part.Nodes)
+
+        ''' Defining linear solver to be used by the variational distance process'''
+        import python_linear_solver_factory #Linear solver for variational distance process
```
```suggestion
        from KratosMultiphysics import python_linear_solver_factory #Linear solver for variational distance process
```
the old way of importing stuff will stop working eventually, please always use the same as when you imports stuff from the CPS app

philbucher commented on applications/CompressiblePotentialFlowApplication/python_scripts/initialize_geometry.py:
```diff
@@ -0,0 +1,252 @@
+
+    def PrintOutput(self,filename):
+        if self.print_output_flag:
+            from gid_output_process import GiDOutputProcess
```
```suggestion
            from KratosMultiphyics.gid_output_process import GiDOutputProcess
```

philbucher commented on applications/CompressiblePotentialFlowApplication/tests/potential_flow_test_factory.py:
```diff
@@ -72,6 +72,15 @@ def test_SmallLiftJumpTest(self):
+        with WorkFolderScope(work_folder):
+            self._runTest(settings_file_name)
+
+            kratos_utilities.DeleteTimeFiles(".")
```
I am curious, does it still write the time files?
I thought I had disabled this globally

marcnunezc replied:
You are right, it does not write it anymore!

philbucher commented on applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp:
```diff
@@ -371,6 +377,9 @@ void DistanceModificationProcess::RecoverOriginalDistance() {
     mModifiedDistancesValues.shrink_to_fit();
+
+    // Restore the TO_SPLIT flag original status
+    this->SetContinuousDistanceToSplitFlag();
```
this is from a different PR right?

marcnunezc replied:
Yes, this is from  #4804!

rubenzorrilla replied:
#4804 already merged. This should disappear merging the current master.

philbucher reviewed (commented):
some comment while browsing through, I might add more comments later

philbucher commented:
btw @marcnunezc can you please [skip the ci](https://github.com/KratosMultiphysics/Kratos/issues/3399) while you are still developing to not block it?
Thanks

inigolcanalejo commented:
Just a small comment I noticed from the previous PR. I believe these lines have the wrong indentation:

https://github.com/KratosMultiphysics/Kratos/blob/d7e5dcc6262bb701dcf8c1820a50247f8525c280/applications/CompressiblePotentialFlowApplication/custom_elements/embedded_incompressible_potential_flow_element.cpp#L87-L99

You can set up VS Code to automatically format your new implementations: https://github.com/KratosMultiphysics/Kratos/wiki/How-to-configure-clang%E2%80%90format

Also, do you really need to allocate aux_matrix? You could directly do:

noalias(rLeftHandSideMatrix) + = prod(DN_DX, trans(DN_DX)) * positive_side_weights(i_gauss);

A question from my side: why are you using GI_GAUSS_2 instead of GI_GAUSS_1 if the elements are linear? Have you tried GI_GAUSS_1? Does it make a difference for your case?

inigolcanalejo commented on applications/CompressiblePotentialFlowApplication/python_scripts/potential_flow_solver.py:
```diff
@@ -55,9 +92,9 @@ def __init__(self, model, custom_settings):
-        # TODO: Create a formulation class helper as soon as there is more than one element is present
-        self.element_name = "IncompressiblePotentialFlowElement"
-        self.condition_name = "PotentialWallCondition"
+        self.embedded_formulation = PotentialFlowFormulation(self.settings["formulation"])
```
Is there a reason why you called this variable embedded_formulation? I would call it just formulation since it can be embedded or not embedded.

rubenzorrilla replied:
Probably the cause is that I called it  `self.embedded_formulation` in the embedded solver. I agree with @inigolcanalejo about calling it `self.formulation`.

Actually, I will change it in the embedded solver just in case there is an eventual derivation from it (even though the proper way to retrieve the formulation name would be to implement a `GetFormulationName()` method).

marcnunezc replied:
Yes, my bad! This was due to copy&pasting the class from the fluid and I forgot to change the name!

inigolcanalejo reviewed (changes requested):
Just another minor comment

marcnunezc committed:
Merge remote-tracking branch 'origin/master' into cps/embedded-pr-2of4

rubenzorrilla commented on applications/CompressiblePotentialFlowApplication/python_scripts/potential_flow_solver.py:
```diff
@@ -68,9 +105,15 @@ def AddVariables(self):
         self.main_model_part.AddNodalSolutionStepVariable(KratosMultiphysics.DISTANCE)
+        self.main_model_part.AddNodalSolutionStepVariable(KratosMultiphysics.DISTANCE_GRADIENT)
+        self.main_model_part.AddNodalSolutionStepVariable(KratosMultiphysics.NODAL_H) # Required for modify_distance_process
+        self.main_model_part.AddNodalSolutionStepVariable(KratosMultiphysics.VELOCITY) # Required for modify_distance_process
```
You need these variables because the modify distance process fixes them to zero in the fully negative and continuous distance case. Am I wrong?

In case you don't need them (I think you need at least the velocity to do the post-process), we should modify the modify distance process to skip this operation in case is not needed since this implies a useless memory overhead. Would you mind open an issue describing the situation?

marcnunezc replied:
Yes, you are right! I will open an issue.

For now we are just using the VELOCITY at the Gauss Points, so actually we dont need to define it as a historical nodal variable.

rubenzorrilla commented on applications/CompressiblePotentialFlowApplication/python_scripts/initialize_geometry.py:
```diff
@@ -0,0 +1,252 @@
+    def CalculateDistance(self):
+        ''' This function calculate the distance to skin for every node in the main_model_part.'''
+        ini_time=time.time()
+        KratosMultiphysics.CalculateDistanceToSkinProcess2D(self.main_model_part, self.skin_model_part).Execute()
```
In the function below (`ExtendDistance`) you allow 3D cases but not in this one. Is this intentionally done?

marcnunezc replied:
Not intentionally, actually I should just restrict it to the 2D cases for now.

rubenzorrilla commented on applications/CompressiblePotentialFlowApplication/python_scripts/initialize_geometry.py:
```diff
@@ -0,0 +1,252 @@
+        raise Exception("expected input shall be a Parameters object, encapsulating a json string")
+    return InitializeGeometryProcess(Model, settings["Parameters"])
+def MoveModelPart(origin, skin_model_part):
+    for node in skin_model_part.Nodes:
```
You should think about migrating this method to c++.

rubenzorrilla commented on applications/CompressiblePotentialFlowApplication/python_scripts/initialize_geometry.py:
```diff
@@ -0,0 +1,252 @@
+    for node in skin_model_part.Nodes:
+        node.X=origin[0]+node.X
+        node.Y=origin[1]+node.Y
+def RotateModelPart(origin, angle, model_part):
```
As well as this one.

marcnunezc committed:
Merge remote-tracking branch 'origin/master' into cps/embedded-pr-2of4

marcnunezc committed:
Merging master

marcnunezc committed:
Fixing compilation error and test

rubenzorrilla commented on applications/CompressiblePotentialFlowApplication/python_scripts/initialize_geometry.py:
```diff
@@ -0,0 +1,252 @@
+        if self.main_model_part.ProcessInfo[KratosMultiphysics.DOMAIN_SIZE] == 2:
+            variational_distance_process = KratosMultiphysics.VariationalDistanceCalculationProcess2D(
+                self.main_model_part,
+                self.linear_solver,
```
If I'm not wrong, you only need the linear solver in here. You can avoid to save it as a member variable if you construct it in here.

rubenzorrilla commented on applications/CompressiblePotentialFlowApplication/python_scripts/initialize_geometry.py:
```diff
@@ -0,0 +1,252 @@
+            In order to perform the refinement, it is needed to calculate the distance gradient, the initial nodal_h and the level_set metric.
+        '''
+        ini_time=time.time()
+        local_gradient = KratosMultiphysics.ComputeNodalGradientProcess2D(self.main_model_part, KratosMultiphysics.DISTANCE,
```
Can you format this. Besides you have the same 2D and 3D issue.

rubenzorrilla commented on applications/CompressiblePotentialFlowApplication/python_scripts/initialize_geometry.py:
```diff
@@ -0,0 +1,252 @@
+    def PrintOutput(self,filename):
+        if self.print_output_flag:
+            from gid_output_process import GiDOutputProcess
+            gid_output = GiDOutputProcess(self.main_model_part,
```
Which is the intention of having the GiDOutputProcess encapsulated in here?. As far as I understand, this process is intended to refine a coarse background mesh according to the level-set function.

marcnunezc replied:
This is used to debug/test cases. It is useful to check the metric values at each step of the remeshing process to check if you would need more or less iterations. Or to check the effect of a good/bad background mesh. 

I can remove this to just use it in case I need it but maybe others can benefit from it when developing other formulations or 3D cases. What would you do?

rubenzorrilla replied:
If it is for debugging it shouldn't be in the master. If you want to keep it for further developments (e.g. 3D) I'd commented so it is not executed by default each time the process is called.

Assuming that each one of the intermediate meshes is saved in a different model part (I don't know if this is the case), you can print them from outside the process by setting a GiD output process for each one of them in the json settings `output_processes` section.

marcnunezc replied:
I understand!
What you are saying about the different model parts could be interesting! I will give it a try.  If i save each intermediate mesh, does it slow in some way the whole simulation?

rubenzorrilla replied:
It will not slow it to much but the memory consumption will significantly increase. But taking into account that you need this for debugging 2D cases (so far) it wouldn't be that much. Of course this is a temporary alternative solution.

rubenzorrilla commented on applications/CompressiblePotentialFlowApplication/python_scripts/initialize_geometry.py:
```diff
@@ -0,0 +1,252 @@
+        node.X = ox+math.cos(angle)*(node.X - ox)-math.sin(angle)*(node.Y - oy)
+        node.Y = oy+math.sin(angle)*(node.X - ox)+math.cos(angle)*(node.Y - oy)
+## All the processes python should be derived from "Process"
+class InitializeGeometryProcess(KratosMultiphysics.Process):
```
I find that the name of this class is misleading and not self-descriptive enough. I'd suggest something like `LevelSetBasedRemeshingProcess`.

By the way, is this compatible with any continuous level-set based formulation?

marcnunezc replied:
> I find that the name of this class is misleading and not self-descriptive enough. I'd suggest something like `LevelSetBasedRemeshingProcess`.

I agree that name is improvable, but your suggestion would be also incomplete. This process, as it is right now, it's used even if there is no remeshing (note that there's a flag that skips the remeshing loop to just initialize the geometry).

```
        self.InitializeSkinModelPart()
        self.CalculateDistance()
        ini_time=time.time()
        while self.step < self.max_iter and self.do_remeshing:
            self.step += 1
            self.ExtendDistance()
            self.RefineMesh()
            self.CalculateDistance()
            self.UpdateParameters()
        self.ModifyFinalDistance()
        self.ExtendDistance()
        self.CopyAndDeleteDefaultDistance()
```
We could separate this loop in three files:
- initialize_skin_to_model_part_process.py (containg move_skin_model_part, rotate_skin_model_part)
- level_set_based_remeshing_process.py
- modify_distance_process.py

Any other suggestion?
> 
> By the way, is this compatible with any continuous level-set based formulation?

Yes, I would say it is!

rubenzorrilla replied:
If what you've done is compatible with any continuous level-set based formulation, it might have sense to separate it in a different utilities, something that would go in line to what @loumalouomega is trying to do in #4558 .

May be it's worth it to discuss it...

marcnunezc replied:
Actually, taking a look at https://github.com/KratosMultiphysics/Kratos/pull/4558#issuecomment-478496561, I'm not sure if what im doing here would work if the skin is moving at each time step. 

@loumalouomega When you can, could you take a look at my scritps? I see you are defining new AnalysisStages to do the remeshing. Is that also for the distance case (when we only refine with resepct to a distance field)?

rubenzorrilla reviewed (commented):
There are some things to be clarified before merging.

philbucher committed:
Update applications/CompressiblePotentialFlowApplication/python_scripts/initialize_geometry.py

Co-Authored-By: marcnunezc <mnunez@cimne.upc.edu>

philbucher committed:
[skip ci] update applications/CompressiblePotentialFlowApplication/python_scripts/initialize_geometry.py

Co-Authored-By: marcnunezc <mnunez@cimne.upc.edu>

marcnunezc committed:
Merge branch 'cps/embedded-pr-2of4' of https://github.com/KratosMultiphysics/Kratos into cps/embedded-pr-2of4

marcnunezc committed:
renaming formulation

marcnunezc committed:
moving linear solver declaration

marcnunezc commented:
Answering @inigolcanalejo 


> Just a small comment I noticed from the previous PR. I believe these lines have the wrong indentation:
> 
> https://github.com/KratosMultiphysics/Kratos/blob/d7e5dcc6262bb701dcf8c1820a50247f8525c280/applications/CompressiblePotentialFlowApplication/custom_elements/embedded_incompressible_potential_flow_element.cpp#L87-L99
> 
> You can set up VS Code to automatically format your new implementations: https://github.com/KratosMultiphysics/Kratos/wiki/How-to-configure-clang%E2%80%90format

Ok thanks!

> Also, do you really need to allocate aux_matrix? You could directly do:
> 
> noalias(rLeftHandSideMatrix) + = prod(DN_DX, trans(DN_DX)) * positive_side_weights(i_gauss);

Not really, it was just to make it cleaner, but I will remove it!

> 
> A question from my side: why are you using GI_GAUSS_2 instead of GI_GAUSS_1 if the elements are linear? Have you tried GI_GAUSS_1? Does it make a difference for your case?

It does not make a difference in my test/cases actually, but just to make sure the "cutting" is not a problem in the solution. Also, I saw that at the FluidDynApp it is set at 2 in the embedded elements!

marcnunezc committed:
Deleting time files remover

marcnunezc committed:
Avoiding auxiliar matrix

philbucher commented:
> > You can set up VS Code to automatically format your new implementations: https://github.com/KratosMultiphysics/Kratos/wiki/How-to-configure-clang%E2%80%90format
> 
> Ok thanks!
> 

actually I would advise against doing this in a regular PR
=> if your format-config files are only slightly different it might happen that the entire file shows up as changed and then reviewing is very difficult. Furthermore it creates conflicts among the devs (this is the reason why it is not done automatically every time a PR is merged)

If you want to format then I recommend doing a separate PR. This I would do in times when not much development is going on since it creates huge conflicts otherwise

rubenzorrilla commented:
> Answering @inigolcanalejo
> 
> > Just a small comment I noticed from the previous PR. I believe these lines have the wrong indentation:
> > https://github.com/KratosMultiphysics/Kratos/blob/d7e5dcc6262bb701dcf8c1820a50247f8525c280/applications/CompressiblePotentialFlowApplication/custom_elements/embedded_incompressible_potential_flow_element.cpp#L87-L99
> > 
> > You can set up VS Code to automatically format your new implementations: https://github.com/KratosMultiphysics/Kratos/wiki/How-to-configure-clang%E2%80%90format
> 
> Ok thanks!
> 
> > Also, do you really need to allocate aux_matrix? You could directly do:
> > noalias(rLeftHandSideMatrix) + = prod(DN_DX, trans(DN_DX)) * positive_side_weights(i_gauss);
> 
> Not really, it was just to make it cleaner, but I will remove it!
> 
> > A question from my side: why are you using GI_GAUSS_2 instead of GI_GAUSS_1 if the elements are linear? Have you tried GI_GAUSS_1? Does it make a difference for your case?
> 
> It does not make a difference in my test/cases actually, but just to make sure the "cutting" is not a problem in the solution. Also, I saw that at the FluidDynApp it is set at 2 in the embedded elements!

In the FluidDynamicsApplication we use order 2 because NS eqs. are non-linear. For the potential flow case, what @inigolcanalejo is pointing makes sense to me. Concerning the cuts, they are always linear, so probably is worth it to try it (this would speed up your simulations quite a lot).

marcnunezc committed:
[skip-ci] formatting

marcnunezc committed:
[skip ci] dummy commit to avoid ci

inigolcanalejo commented:
Is this ready to be merged? Shall I approve or is there still some discussion going on?

marcnunezc commented:
There's still some discussion going on! I will comment again once it's ready!

inigolcanalejo commented:
Alright! I saw that you did some changes and I just wanted to make sure I am not blocking.

marcnunezc committed:
Merge remote-tracking branch 'origin/master' into cps/embedded-pr-2of4

marcnunezc committed:
updating parameter name

marcnunezc committed:
Merge remote-tracking branch 'origin/master' into cps/embedded-pr-2of4

marcnunezc committed:
adapting test to master changes

marcnunezc committed:
reducing number of gauss points in the embedded incompr. element

marcnunezc committed:
removing pring output

marcnunezc committed:
adding file for move_movel_part process

marcnunezc committed:
removing print output function

marcnunezc committed:
adding move model part process to python

marcnunezc committed:
Using move-model-aprt process and defining protected methods

marcnunezc committed:
renaming process

marcnunezc committed:
renaming classes and test

marcnunezc committed:
adding auxiliar variable to rotate model part

marcnunezc committed:
adding test for the new process

marcnunezc commented:
Hi! 
This branch is ready to final review.

@rubenzorrilla and @inigolcanalejo , I implemented your comments including:
-Using 1 gauss points in the element
-Renaming formulation in the solver
-Creating  a cpp process to do the skin_model_part movements/rotations
-Renaming of the process to level_set_remeshing_process
-Removing gid_output calls inside the process

I added a cpp test for this new process.

Let me know if you have more comments!

Regarding the discussion on adding this to #4558 , @loumalouomega recommended me to keep this script here for now, since the remeshing scripts in #4558  are aimed for error-based remeshing, which requires to solve twice the system (in my scripts im remeshing in terms of the distance, no need to solve the problem beforehand). I would suggest to keep the script here for now and once the whole embedded workflow is implemented we can discuss to move it elsewhere (since steps 3 or 4 of this PRs might force changes in the script).

rubenzorrilla commented on applications/CompressiblePotentialFlowApplication/custom_processes/move_model_part_process.h:
```diff
@@ -0,0 +1,187 @@
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
```
Header format

rubenzorrilla commented on applications/CompressiblePotentialFlowApplication/custom_processes/move_model_part_process.h:
```diff
@@ -0,0 +1,187 @@
+    }
+
+    /// Destructor.
+    ~MoveModelPartProcess() override {}
```
```suggestion
    ~MoveModelPartProcess() = default;
```

rubenzorrilla commented on applications/CompressiblePotentialFlowApplication/custom_processes/move_model_part_process.h:
```diff
@@ -0,0 +1,187 @@
+        KRATOS_TRY;
+
+        for(auto it_node=mrModelPart.NodesBegin(); it_node!=mrModelPart.NodesEnd(); ++it_node){
+            array_1d<double, 3>& r_coordinates = it_node->Coordinates();
```
```suggestion
            auto &r_coordinates = it_node->Coordinates();
```

rubenzorrilla commented on applications/CompressiblePotentialFlowApplication/custom_processes/move_model_part_process.h:
```diff
@@ -0,0 +1,187 @@
+        for(auto it_node=mrModelPart.NodesBegin(); it_node!=mrModelPart.NodesEnd(); ++it_node){
+            array_1d<double, 3>& r_coordinates = it_node->Coordinates();
+
+            for (std::size_t i_dim = 0; i_dim<mOrigin.size();i_dim++){
```
I think that `mOrigin.size()` is always 3. Am I wrong?

marcnunezc replied:
Yes it is!

philbucher commented:
I saw some things while browsing through but right now I have no time to write it all down. If you want to wait a few days I will add my comments

rubenzorrilla commented on applications/CompressiblePotentialFlowApplication/custom_processes/move_model_part_process.h:
```diff
@@ -0,0 +1,187 @@
+    void Execute() override
+    {
+        KRATOS_TRY;
+
```
```suggestion
#pragma omp parallel for
for(int i_node = 0; i_node = static_cast<int>(mrModelPart.NumberOfNodes()); ++i_node){
    auto it_node = mrModelPart.NodesBegin() + i_node;
```

rubenzorrilla commented on applications/CompressiblePotentialFlowApplication/custom_processes/move_model_part_process.h:
```diff
@@ -0,0 +1,187 @@
+    Vector mRotationPoint;
+    double mRotationAngle;
+    double mSizingMultiplier;
+    Parameters mrOptions;
```
I think that you don't need this one if you save all the parameters entries.

rubenzorrilla commented on applications/CompressiblePotentialFlowApplication/custom_processes/move_model_part_process.h:
```diff
@@ -0,0 +1,187 @@
+    MoveModelPartProcess& operator=(MoveModelPartProcess const& rOther);
+
+    /// Copy constructor.
+    MoveModelPartProcess(MoveModelPartProcess const& rOther);
```
```suggestion
    MoveModelPartProcess(MoveModelPartProcess const& rOther) = delete;
```

rubenzorrilla replied:
And you can move it to the life cycle section.

rubenzorrilla commented on applications/CompressiblePotentialFlowApplication/custom_processes/move_model_part_process.h:
```diff
@@ -0,0 +1,187 @@
+    ///@{
+
+    /// Assignment operator.
+    MoveModelPartProcess& operator=(MoveModelPartProcess const& rOther);
```
```suggestion
    MoveModelPartProcess& operator=(MoveModelPartProcess const& rOther) = delete;
```

rubenzorrilla replied:
And you can move it to the life cycle section.

rubenzorrilla commented on applications/CompressiblePotentialFlowApplication/custom_processes/move_model_part_process.h:
```diff
@@ -0,0 +1,187 @@
+//    |  /           |
```
I think that there is no reason for this to be header only.

rubenzorrilla commented on applications/CompressiblePotentialFlowApplication/tests/cpp_tests/test_processes.cpp:
```diff
@@ -0,0 +1,65 @@
+            "sizing_multiplier"             : 2.0
+
+        })" );
+      double pi = 3.1415926535897;
```
I think that @loumalouomega added Pi as a Kratos global variable.

rubenzorrilla commented:
BTW, have a look to Codacy output.

marcnunezc committed:
separating move_model part into h and cpp

marcnunezc committed:
using global pi variable

marcnunezc committed:
using openmp in the process

marcnunezc committed:
codacy suggestions

marcnunezc commented:
@rubenzorrilla suggestions are implemented! Do you have more suggestions after ruben's comments @philbucher?

Regards!

philbucher commented on applications/CompressiblePotentialFlowApplication/python_scripts/potential_flow_solver.py:
```diff
@@ -68,9 +105,15 @@ def AddVariables(self):
         self.main_model_part.AddNodalSolutionStepVariable(KCPFApp.VELOCITY_POTENTIAL)
         self.main_model_part.AddNodalSolutionStepVariable(KCPFApp.AUXILIARY_VELOCITY_POTENTIAL)
-
+        # Embedded variables
```
those variables are only used in the embedded case?

I am a astonished of how many historical variables you need
Have you thought abt splitting the solvers in the same fashion as it is done in FluidDynamics?
This way there would be no overhead for the body-fitted case

@rubenzorrilla do all those variables (and their usage in the various processes) really have to be historical?
This seems like quite an overhead to me

marcnunezc replied:
Yes, the GEOMETRY_DISTANCE is only for the embedded case!

The rest of variables are defined as historical since they are either used or required in/by Kratos processes. 

DISTANCE, DISTANCE_GRADIENT in calculate_distance_to_skin_process, metrics_level_set_process
FLAG_VARIABLE in variational_distance_process
NODAL_H, VELOCITY, PRESSURE in modify_distance_process (see #4834)

rubenzorrilla replied:
They shouldn't, but right now the level-set calculation processes assume that the `DISTANCE` variables are historical. Same happens for the distance modification ones, which require historical `NODAL_H`, `VELOCITY` and `PRESSURE`. I think that @marcnunezc already opened an issue complaining about that.

philbucher commented on applications/CompressiblePotentialFlowApplication/tests/test_CompressiblePotentialFlowApplication.py:
```diff
@@ -32,7 +32,7 @@ def AssembleTestSuites():
     smallSuite = suites['small']
     smallSuite.addTest(PotentialFlowTests('test_Naca0012SmallAdjoint'))
-    #smallSuite.addTests(KratosUnittest.TestLoader().loadTestsFromTestCases([TestPotentialElement2D]))
+    smallSuite.addTest(PotentialFlowTests('test_EmbeddedCircleNoWake'))
```
check e.g. [here](https://github.com/KratosMultiphysics/Kratos/blob/master/applications/StructuralMechanicsApplication/tests/test_StructuralMechanicsApplication.py#L389) how to automatically add all tests from a test case, this way you will never forget to add it

=> I see you had this already (commented), I highly recommend using it. A test that is not run is kinda useless ;)

marcnunezc replied:
Can you elaborate what does that line do? Right now if i run the script `test_CompressiblePotentialFlowApplication.py` the 'Naca0012' test and 'Embedded' test run correctly, along the cpp_tests.

(this is without the test loader)

philbucher replied:
it detects all tests that are in the test-class => meaning that you don't have to add every single test manually

imagine you write a new test, e.g. `test_EmbeddedCircleNoWake_2`, but you forget to add it to the smallsuite, then it will never be executed
This won't happen if you use `loadTestsFromTestCases`, which does what I explained

marcnunezc replied:
Doing the following (note that the test addition is commented):
```
    smallSuite = suites['small']
    # smallSuite.addTest(PotentialFlowTests('test_Naca0012SmallAdjoint'))
    # smallSuite.addTest(PotentialFlowTests('test_EmbeddedCircleNoWake'))
    KratosUnittest.TestLoader().loadTestsFromTestCases([PotentialFlowTests])
```

Seems to do nothing (no tests are run). Am I required to add every specific test regardless?

philbucher replied:
yes, you are not adding the tests to the smallsuite
=> `smallSuite.addTests(KratosUnittest.TestLoader().loadTestsFromTestCases([PotentialFlowTestsement]))` should work

marcnunezc replied:
Ok now it works! Thanks!

philbucher replied:
if in doubt check how it is done in StructuralMechanics, there it is used exensively

philbucher commented on applications/CompressiblePotentialFlowApplication/custom_processes/move_model_part_process.h:
```diff
@@ -0,0 +1,91 @@
+namespace Kratos
+{
+
+class MoveModelPartProcess: public Process
```
I think throughout Kratos many of those exist, maybe eventually we can have this unified in the Core

I also have a version of this (works also in 3D), but only python, haven't done the effort yet to port it to Kratos...

Just saying that hopefully some day we will have a global solution :)

marcnunezc committed:
using test loader

marcnunezc committed:
Merge remote-tracking branch 'origin/master' into cps/embedded-pr-2of4

marcnunezc commented:
Hi! This PR is ready to merge! Could someone approve it if there are no more comments?

Thanks!

rubenzorrilla commented:
I've no objection to this to be merged. I let @philbucher to do the final approve.

marcnunezc committed:
Merge remote-tracking branch 'origin/master' into cps/embedded-pr-2of4

marcnunezc committed:
removing extend distance, resetting metric value and reverting test values

marcnunezc committed:
removing unused typedefs

marcnunezc commented:
I did a small change in which I reset the metric_tensor to 0 at every step in case more than one step is performed in the remeshing. Also I removed the "extension" of the distance field using the variational_distance_process at the very last step of the remeshing script.

The branch is ready to be merged regardless!

philbucher reviewed (dismissed):
Sorry did not see my review was missing 

Good job 👍🏻

marcnunezc commented:
@inigolcanalejo seems like I also need your approval since you requested changes!

Could you also approve? Thanks!

inigolcanalejo commented:
Sorry, I also didn't see that my review was missing.

marcnunezc committed:
dummy commit to restart appveyor

marcnunezc commented:
Hi I added a dummy commit to restart Appveyor since it was stuck in "Expected". Could someone reapprove? Sorry for the inconvenience

Is there another way to restart the checks that we can use?

philbucher reviewed (approved):
@marcnunezc if you only merge master you won't need reapproval (unless there are conflicts)
